### PR TITLE
Add transformation to indexing intersection of homomorphic mapepd type and generic parameter

### LINKED ISSUE #25181: SpyObj<T>[keyof T] not resolving to the type SpyObj members (Jasmine typings)
**TypeScript Version:**  2.9.2

**Code**

```ts
interface Spy {
  (...params: any[]): any;

  identity: string;
  and: Function;
  mostRecentCall: {args: any[];};
  argsForCall: any[];
}

type SpyObj<T> = T&{
  [k in keyof T]: Spy;
}

declare function createSpyObj<T>(
    name: string, names: Array<keyof T>): SpyObj<T>;

/**
 * Creates a jasmine mock where every given method returns the spy object.
 */
function mock<T>(spyName: string, methodNames: Array<keyof T>): SpyObj<T> {
  const spyObj = createSpyObj<T>(spyName, methodNames);
  for (const methodName of methodNames) {
    spyObj[methodName].and.returnValue(1);  // fails, Property 'and' does not exist on type 'SpyObj<T>[keyof T]'.
  }
  return spyObj;
}
```

**Expected behavior:**

In TS2.8, `spyObj[methodName]` above was any, allowing `and` to be referenced on it.  Ideally, this would figure out that all members of `spyObj` must at least be a `Spy`.

**Actual behavior:**

In TS 2.9, there's a type `SpyObj<T>[keyof T]`, but apparently it does not resolve to `Spy`?

**Playground Link:** 

http://www.typescriptlang.org/play/#src=%0Ainterface%20Spy%20%7B%0A%20%20(...params%3A%20any%5B%5D)%3A%20any%3B%0A%0A%20%20identity%3A%20string%3B%0A%20%20and%3A%20Function%3B%0A%20%20mostRecentCall%3A%20%7Bargs%3A%20any%5B%5D%3B%7D%3B%0A%20%20argsForCall%3A%20any%5B%5D%3B%0A%7D%0A%0Atype%20SpyObj%3CT%3E%20%3D%20T%26%7B%0A%20%20%5Bk%20in%20keyof%20T%5D%3A%20Spy%3B%0A%7D%0A%0Adeclare%20function%20createSpyObj%3CT%3E(%0A%20%20%20%20name%3A%20string%2C%20names%3A%20Array%3Ckeyof%20T%3E)%3A%20SpyObj%3CT%3E%3B%0A%0A%2F**%0A%20*%20Creates%20a%20jasmine%20mock%20where%20every%20given%20method%20returns%20the%20spy%20object.%0A%20*%2F%0Afunction%20mock%3CT%3E(spyName%3A%20string%2C%20methodNames%3A%20Array%3Ckeyof%20T%3E)%3A%20SpyObj%3CT%3E%20%7B%0A%20%20const%20spy%20%3D%20createSpyObj%3CT%3E(spyName%2C%20methodNames)%3B%0A%20%20for%20(const%20methodName%20of%20methodNames)%20%7B%0A%20%20%20%20spy%5BmethodName%5D.and.returnValue(spy)%3B%0A%20%20%7D%0A%20%20return%20spy%3B%0A%7D%0A

### PR DESCRIPTION
**Pull request by weswigham**
In `getSimplifiedType`, adds a branch which transforms `(T & {[P in keyof T]: Foo})[K]` (which we can't reason about, thanks to how we combine the intersection before looking for members on its apparent type) into `{[P in keyof T]: T[P] & Foo}[K]` (which we can reason about).

Specifically, in the first case, when calculating its constraint, we see that it's an indexed access, attempt to simplify it (and it doesn't), then we get the base constraints of `T & {[P in keyof T]: Foo}` and `K`, and produce a new indexed access - that becomes `({} & {})[string | number | symbol]`, which then becomes the `error` type.

Now, when we attempt to simplify it, we apply the above mapping, and the index type `{[P in keyof T]: T[P] & Foo}[K]` in the provided sample simplifies to `T[K] & Foo` (since `K` _is_ `keyof T`) whose constraint is simply `Foo`.

Fixes #25181

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -9042,6 +9050,33 @@ namespace ts {
...
+                            const match = forEach((<IntersectionType>objectType).types, (t, i) => getIndexType(t) === constraint ? (matchIndex = i, t) : undefined)!;
+                            const indexedMatch = getIndexedAccessType(match, getTypeParameterFromMappedType(t as MappedType));
+                            const newTemaplate = getIntersectionType([indexedMatch, getTemplateTypeFromMappedType(t as MappedType)]);
+                            const newMappedType = <MappedType>createObjectType(ObjectFlags.Mapped, t.symbol);
+                            newMappedType.declaration = (t as MappedType).declaration;
+                            newMappedType.templateType = newTemaplate;
+                            newMappedType.typeParameter = (t as MappedType).typeParameter;
+                            newMappedType.constraintType = (t as MappedType).constraintType;
+                            const newSet = (<IntersectionType>objectType).types.slice();
+                            newSet.splice(matchIndex!, 1); // Remove the generic
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: `orderedRemoveItemAt` (or `unorderedRemoveItemAt` if it doesn't matter)

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -9042,6 +9050,33 @@ namespace ts {
...
+                            newSet.splice(newSet.indexOf(t), 1, newMappedType); // Remove the original mapped type
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Why not just index at the original location?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -9042,6 +9050,33 @@ namespace ts {
...
+                // transformed type that moves the top-level intersection into the mapped type: '{[P in keyof T]: T[P] & Foo}[K]'
+                // If we don't do such a transformation now, due to how we reason over intersections, we will never come to the
+                // realization that the member we're trying to pluck out is influenced by both `T[P]` _and_ the mapped type template
+                const isInSet = isMappedTypeOverKeyofGenericInSet((<IntersectionType>objectType).types);
+                if (some((<IntersectionType>objectType).types, isInSet)) {
+                    for (const t of (<IntersectionType>objectType).types) {
+                        if (isInSet(t)) {
+                            const constraint = getConstraintTypeFromMappedType(t as MappedType);
+                            let matchIndex: number | undefined;
+                            const match = forEach((<IntersectionType>objectType).types, (t, i) => getIndexType(t) === constraint ? (matchIndex = i, t) : undefined)!;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Extract `getIndexType` out so you only compute it once.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -9042,6 +9050,33 @@ namespace ts {
...
                     return type.simplified = getSimplifiedType(getIndexedAccessType(getIntersectionType(nonNeverTypes), type.indexType));
                 }
+
+                // Given an indexed access type T[K], if T is an intersection containing a generic type and one or
+                // more mapped types which map over the keys of that generic, '(T & {[P in keyof T]: Foo})[K]', return a
+                // transformed type that moves the top-level intersection into the mapped type: '{[P in keyof T]: T[P] & Foo}[K]'
+                // If we don't do such a transformation now, due to how we reason over intersections, we will never come to the
+                // realization that the member we're trying to pluck out is influenced by both `T[P]` _and_ the mapped type template
+                const isInSet = isMappedTypeOverKeyofGenericInSet((<IntersectionType>objectType).types);
+                if (some((<IntersectionType>objectType).types, isInSet)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Could you turn this into a `findIndex` and then avoid the inner loop?

File: src/compiler/checker.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/checker.ts]: ... fair point. I was in a splicy mood, I suppose.

**Comment by weswigham** [comment]
Superseded by #26281, which is much more elegant, IMO.

**Pull request closed without merging**